docs/design/declaring_entities.md

```diff
@@ -0,0 +1,233 @@
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [Matching redeclarations of an entity](#matching-redeclarations-of-an-entity)
+    -   [Details](#details)
+-   [`extern` and `extern library`](#extern-and-extern-library)
+    -   [Effect on indirect imports](#effect-on-indirect-imports)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+Entities may have up to three declarations:
+
+-   An optional, owning forward declaration.
+    -   For example, `class MyClass;`.
+    -   This must come before the definition. The API file is considered to be
+        before the implementation file.
+-   A required, owning definition.
+    -   For example, `class MyClass { ... }`.
+    -   The definition might be the _only_ declaration.
+-   An optional, non-owning `extern library "<owning_library>"` declaration.
+    -   For example, `extern library "OtherLibrary" class MyClass;`.
+    -   It must be in a separate library from the definition.
+    -   The owning library's API file must import the `extern` declaration, and
+        must also contain a declaration.
+    -   The owning library's declarations must have the `extern` modifier
+        (without `library`).
+        -   For example, `extern class MyClass;`.
+
+For example, a library can have a forward declaration of an entity in the API
+file, and use the implementation file for the entity's definition. Putting the
+definition in an implementation file this way can reduce the dependencies for
+API file evaluation, improving compile time. This is commonly done with
+functions. For example:
+
+```
+library "MyLibrary";
+
+fn DoSomething();
+```
+
+```
+impl library "MyLibrary";
+
+fn DoSomething {
+  ...
+}
+```
+
+## Matching redeclarations of an entity
+
+In order to determine whether two redeclarations refer to the same entity, we
+apply the rules:
+
+-   Two named declarations _declare the same entity_ if they have the same scope
+    and the same name.
+-   One declaration _redeclares_ another if they declare the same entity and the
+    second declaration appears after the first, including the case where the
+    first declaration is imported. In this case, the second declaration is said
+    to be a _redeclaration_ of the first.
+-   Two declarations _differ_ if the sequence of tokens in the declaration
+    following the introducer keyword and the optional scope, up to the semicolon
+    or open brace, is different, except for `unused` modifiers on parameters.
+-   The program is invalid if it contains two owning declarations of the same
+    entity that differ.
+    -   The non-owned `extern library` declarations will only use semantic
+        matching for redeclarations, not syntactic matching.
+
+```carbon
+class A {
+  fn F(n: i32);
+  fn G(n: i32);
+}
+
+// ❌ Invalid: The parameter name differs.
+fn A.F(m: i32) {}
+
+// ❌ Invalid: The parameter type differs syntactically.
+fn A.G(n: (i32)) {}
```

Comment:
Would be nice to show more examples.

```suggestion
fn A.G(n: (i32)) {}

// ✅ Valid: only difference is the scope is explicit.
fn A.F(n: i32) {} 
```

Possibly also an example showing an extern library declaration with semantic but not syntactic matching?

Reply:
Added your example, but rephrasing why this is valid. Also just dropping the switch to `G`, it feels odd when there's a second `F` example and I'm hesitant to add an `H`.

This is currently a short summary, and `extern library` is a peripheral feature, so I think it should probably be kept for details.

I see why you think this isn't syntactic matching -- but I think this is just a problem with how syntactic matching was summarized in #3762, it wasn't getting into detail for qualified declarations. That is, for each scope and the entity itself (each of `A` and `F(n: i32)` here), the entity name and parameters should syntactically match; as you point out, it's inaccurate to say "following the introducer keyword".

Rather than trying to address that here, I'm eliding that for details of syntactic versus semantic matching (the original proposal has a section for this, but it didn't feel like it should be copy-pasted).